### ISSUE TITLE
Pan timeline on customTime drag

### ISSUE THREAD

**Issue by aytech** [almende/vis#657]
Hello,
When I enable showCustomTime for the timeline, I get the vertical bar that I can move across the visible range on the timeline. And here's the situation I'm faced with now: if I want to move the bar beyond the visible range, I have to drag the bar to the far end, release it, grab and move the timeline, then grab the vertical bar again. Then repeat ...
I have set up an example where I can detect the offset from the end/start time of the timeline axis and move the timeline as the customTime hits that offset. The example can be seen here: http://jsfiddle.net/aytech/mn8ck81p/14/
My main problem with this approach is that I lose the vertical bar when the timeline is moved since the bar is centered on the timeline on moveTo. I thought that it would be useful to have an option to move the timeline continuously (animated), while the customTime bar is within the offset area (say, 1/5 of the timeline width on both sides). The customTime vertical bar would remain under the mouse cursor. If the bar is moved outside of the offset area, the timeline animation would stop.

Is there a similar functionality in the timeline? I couldn't find anything in the docs, maybe I missed something...
If not, I'd like to try to implement such a feature.

Thanks,
Oleg

**Comment by AlexDM0** [comment]
Hi Oleg,

Why not zoom out as you drag? 

Regards,

Alex

**Comment by aytech** [comment]
Hi Alex,
You mean to update the start/end dates on the fly? I didn't see a method to zoom in/out...

Oleg

**Comment by AlexDM0** [comment]
Hi Oleg,

No I ment, grab it with the mouse, the scroll out with the mousewheel :)

Cheers

**Comment by aytech** [comment]
Alex,
Nah, just tried it in that sandbox on the URL above, first of all, you loose the vertical bar as you zoom out, all the events get jammed together (those dummy icons I put there), and if you try to zoom in and zoom out couple times, you get lost completely, dates are changing all the time and it becomes confusing.
Plus, you can't expect an ordinary user to operate left click and scroll at the same time :)
The way I set it up in that sandbox works, but the bar centering ruins it... I was thinking to call moveTo in a loop, but looks like there's some kind of easing there, so the effect won't be nice....

Any other suggestions? :)

Oleg

**Comment by josdejong** [comment]
I guess what you mean is the same as how all browsers and editors work: when you drag selected text up/down, the page will automatically scroll further up/down when you reach the edge of the window.

This same mechanism will be very useful for the Timeline I think, both when dragging the custom time bar as well as dragging items.

**Comment by aytech** [comment]
Yes, it could be the same type of animation as the one used in moveTo method, just without easing. When a bar is moved too close to the left or right, the timeline would start moving and would keep moving unless the bar is moved out of the "trigger zone", which could be defined, say at 5% of the timeline width on both sides.
This could be set by an option or enabled by default and ability to disable via an option

Regards,
Oleg

**Comment by AlexDM0** [comment]
So essentially, all that is needed is that the drag events are updated on a window change. That would mean your solution in the fiddle works, the zooming in and out works and everyone could implement their own solution the way they like.

I think I'd prefer to have it solved like that instead of a predefined option with a trigger zone. To make it easier for people we could show this behaviour and the appropriate functions in an example.

What do you think?

Regards,

Alex

**Comment by josdejong** [comment]
Indeed drag events should be updated on a window change.

I don't agree Alex, I think a "trigger zone" on the left and right side of the Timeline will work very nicely to drag items and the time bar further than the edge of the Timeline. And this works as people expect it to work. I don't think you're going to scroll while you're dragging, especially not on a touch device...

I agree with @aytech that we need this trigger zone's, which let the Timeline start moving at a certain speed. The width of these zones will be configurable automatically using css. We could make the speed of the timeline movement configurable too, but that may not even be needed at all if we have good default behavior.

**Comment by aytech** [comment]
I did a bit of a research on other timelines out there, none of them have that kind of functionality, at least the ones I found, mostly the solution is to drag the whole timeline. Since vis.js has the custom vertical bar, one would expect that there would be an option to drag it beyond the timeline axis without too many clicks.
I read about the feature-obesity caution on the vis webpage, that's why I suggested to have it enabled by default and disabled by an option if necessary.

Oleg

**Comment by aytech** [comment]
Hi guys,
Any resolution? :) I looked into the code and it looks like this will not be an easy task at all. I tried about 3 solutions using the existing functionality in vis.js and scrapped them all, biggest problem was to move the timeline smoothly and consistently while not loosing the mousedown event on the CustomTime element...

Oleg

**Comment by AlexDM0** [comment]
Hi Oleg,

I think you'd need to work with (or around) the deltaX variable in the onDrag of the customTime. There needs to be a way to update that on a rangeChange event. This should solve the alignment issue when the range changes. This should also be done for items. 

I'm not sure what you mean by losing the event? In your initial example all I can tell is that the triggerzone is not updated on rangechange. The dragging still works but it is offset, which should be solved by my suggestion above.  

I'd suggest you insert a requestAnimationFrame with a increment in the start and end time (depending on back or forth of course). This increment can be based linearly or exponentially how far the pointer is in the triggerzone. This change should automatically trigger a rangeChange event which you can use to update the triggerzone and the deltax.

Regards,

Alex

**Comment by aytech** [comment]
Hi Alex,
Thanks for the tip, I will try to dive deeper into the idea. 
By loosing the event I mean that the moment I add customTime object to the timeline "components" collection, it kind of "sticks" to the axis, so that when you drag the timeline, it moves along with it. When I move the timeline with the Range setRange method (used in moveTo), the customTime bar and the mouse cursor become unsynchronized. I tried to detach the customTime object from the timeline on drag and reattach it on the drag end, but then I lost the cursor, and the customTime line had to be clicked again in order to be moved further. 

As for the items, I guess the timeline's (Core) redraw function will do the job after the range change, am I right? 

Thanks,
Oleg

**Comment by AlexDM0** [comment]
Hi Oleg,

Maybe @josdejong could clarify this a bit since he built the timeline and has a better understanding of the codebase. From what I can tell, the only real problem is the deltaX in the onDrag handler that is not corrected for the change in the range. The smooth scrolling can be done using a number of different methods once that is solved. The event propagation should handle the rest.

If the ondrag function is extended iwth the triggerarea logic then that could all be updated on a rangechange event.

Regards,

Alex

**Comment by aytech** [comment]
You mean the onDrag handler for the customTime bar? I believe there's no direct reference to the timeline object inside the customTime. If I'm to update the deltaX logic in the customTime, I'd have to listen to the rangechange event inside the onDrag method of the customTime, right?

**Comment by AlexDM0** [comment]
> You mean the onDrag handler for the customTime bar

Yes initially to have a working prove of concept. After that maybe we can find a way that inserts this into all drag handlers of draggable items. I assume the triggerzone logic would be the same for all of them. 

>If I'm to update the deltaX logic in the customTime, I'd have to listen to the rangechange event inside the onDrag method of the customTime, right?

Not necessarily, I'd recommend that you mimic the deltaX variable using a this. variable of the customTime object. Then you can register the function that listens for the event in the constructor. There should be an event emitter in the body object that is passed to the constructor that can be used for intermodule communication. 

Regards,

Alex

**Comment by josdejong** [comment]
@aytech I will have a look into it, I don't know by heart what's causing the problem here. The main problem here is that the positioning of the dragged customTime bar is not corrected when calling setWindow while dragging. Shall I figure this out and try to fix it? I expect to be able to spend time on this tomorrow and beginning of next week.

With that out of the way, your own solution should work correctly. Next step is to implement this in the Timeline itself, both for dragging items as well as the custom time bar.

**Comment by aytech** [comment]
@josdejong: If you could, that'd be awesome. I'd be happy to help, too, so if you'll have something I could help you with, don't hesitate to ask :)

@AlexDM0: thanks for the tip, I guess I'll hold onto that thought for now, will try to tinker in the forked repo to see how it all comes together, yesterday I almost got my brain melting trying to figure out all the dependencies and logic, and got that bug in my head that I'm close to something :-)

**Comment by aytech** [comment]
Hi @josdejong: do you think the proposed functionality is doable?

**Comment by josdejong** [comment]
Yes, sorry it's not yet fixed. I'm not sure when I can make time for it.

**Comment by aytech** [comment]
OK, I'll try to make some time for it too, see what I can do. Is it all right to keep this issue open for a while?

**Comment by josdejong** [comment]
Sure, lets keep the issue open until resolved.

**Comment by sbussetti** [comment]
+1 on this.  Same issue exists for dragging items around.  Even if I change the position of the item / customTime at the same time that I am moving the window around, the item immediately moves back  to it's previous position -- I assume because of this issue where the internal drag handler is not getting adjusted if the window is moved during a drag operation.  Any update?

**Comment by tatsujb** [comment]
does anyone have a more or less working hack? I can't get it to work.
